### PR TITLE
[bitnami/postgres-ha] feat: add logical backup to pg data

### PR DESCRIPTION
**Pull request by abarrak**
### Description of the change

This mostly imports of the backup feature from [the main pg chart](https://github.com/bitnami/charts/tree/main/bitnami/postgresql#backup-parameters) with changes required for pg-ha.

### Benefits

Enable the logical backup for the database

### Possible drawbacks

None. The cronjob would connect directly the PG instance svc as HOST with the postgres user per configurations.

### Applicable issues

N/A.

### Additional information

Testing (Verification):

- test-values.yaml
```yaml
backup:
  ## @param backup.enabled Enable the logical dump of the database "regularly"
  enabled: true
  cronjob:
    schedule: "36 13 * * *"
    storage:
      resourcePolicy: "keep"
```

##### Running the chart with backup enabled:

```console
$ helm install -n postgres postgres -f test-values.yaml ./ 
NAME: postgres
LAST DEPLOYED: Tue Oct 31 16:40:31 2023
NAMESPACE: postgres
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
CHART NAME: postgresql-ha
CHART VERSION: 12.0.4
APP VERSION: 16.0.0
** Please be patient while the chart is being deployed **
PostgreSQL can be accessed through Pgpool via port 5432 on the following DNS name from within your cluster:

    postgres-postgresql-ha-pgpool.postgres.svc.cluster.local

Pgpool acts as a load balancer for PostgreSQL and forward read/write connections to the primary node while read-only connections are forwarded to standby nodes.

To get the password for "postgres" run:

    export POSTGRES_PASSWORD=$(kubectl get secret --namespace postgres postgres-postgresql-ha-postgresql -o jsonpath="{.data.password}" | base64 -d)

To get the password for "repmgr" run:

    export REPMGR_PASSWORD=$(kubectl get secret --namespace postgres postgres-postgresql-ha-postgresql -o jsonpath="{.data.repmgr-password}" | base64 -d)

To connect to your database run the following command:

    kubectl run postgres-postgresql-ha-client --rm --tty -i --restart='Never' --namespace postgres --image docker.io/bitnami/postgresql-repmgr:16.0.0-debian-11-r11 --env="PGPASSWORD=$POSTGRES_PASSWORD"  \
        --command -- psql -h postgres-postgresql-ha-pgpool -p 5432 -U postgres -d postgres

To connect to your database from outside the cluster execute the following commands:

    kubectl port-forward --namespace postgres svc/postgres-postgresql-ha-pgpool 5432:5432 &
    psql -h 127.0.0.1 -p 5432 -U postgres -d postgres
```

#### Backup Verification:

```console
$ k get cronjobs
NAME                                          SCHEDULE      SUSPEND   ACTIVE   LAST SCHEDULE   AGE
postgres-postgresql-ha-postgresql-pgdumpall   42 13 * * *   False     0        <none>          50s

$ k get pvc
NAME                                          STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
data-postgres-postgresql-ha-postgresql-0      Bound    pvc-66203911-098f-4f6f-97c2-8a958d1a64cb   8Gi        RWO            standard       61s
data-postgres-postgresql-ha-postgresql-1      Bound    pvc-cb08920f-cce0-4ebe-adfa-963fbc3ddf80   8Gi        RWO            standard       61s
data-postgres-postgresql-ha-postgresql-2      Bound    pvc-1c9b349e-3233-4008-94c3-267ad5daa4c0   8Gi        RWO            standard       61s
postgres-postgresql-ha-postgresql-pgdumpall   Bound    pvc-fdba2341-1c35-4714-bbc6-978a1f9ec0aa   8Gi        RWO            standard       61s

$ k logs postgres-postgresql-ha-postgresql-pgdumpall-28312662-lrdcv
pg_dump: warning: there are circular foreign-key constraints on this table:
pg_dump: detail: nodes
pg_dump: hint: You might not be able to restore the dump without using --disable-triggers or temporarily dropping the constraints.
pg_dump: hint: Consider using a full dump instead of a --data-only dump to avoid this problem.

$ ls -al
total 20
drwxrwxrwx 2 root root 4096 Oct 31 13:42 .
drwxr-xr-x 3 root root 4096 Oct 31 13:54 ..
-rw-r--r-- 1 1001 root 9894 Oct 31 13:42 pg_dumpall-2023-10-31-13-42.pgdump
```

### Checklist

- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [x] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami-labs/readme-generator-for-helm)
- [x] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [x] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: bitnami/postgresql-ha/README.md [hunk 1]
```diff
@@ -567,26 +567,63 @@ Additionally, if `persistence.resourcePolicy` is set to `keep`, you should manua
...
-| `service.sessionAffinity`                             | Control where client requests go, to the same pod or round-robin                              | `None`       |
-| `service.sessionAffinityConfig`                       | Additional settings for the sessionAffinity                                                   | `{}`         |
-| `service.annotations`                                 | Provide any additional annotations both for PostgreSQL and Pgpool services                    | `{}`         |
-| `service.serviceLabels`                               | Labels for PostgreSQL service                                                                 | `{}`         |
-| `service.headless.annotations`                        | Annotations for the headless service.                                                         | `{}`         |
-| `networkPolicy.enabled`                               | Enable NetworkPolicy                                                                          | `false`      |
-| `networkPolicy.allowExternal`                         | Don't require client label for connections                                                    | `true`       |
```
Comment (abarrak) [inline bitnami/postgresql-ha/README.md:587]: not sure why bot cuts off networkPolicy and append into the new table.

**Review by aoterolorenzo — CHANGES_REQUESTED** [review]
Hi @abarrak ,

Thanks for the contribution!

Seems the [goss tests](https://github.com/bitnami/charts/tree/main/.vib/postgresql-ha/goss) are failing here. Please mind that all necessary parameters should be present in the [runtime_parameters.yaml](https://github.com/bitnami/charts/blob/main/.vib/postgresql-ha/runtime-parameters.yaml) in order to properly run the tests. This feature should not be tested IMO, so actually shouldn't be failing since it would take all default new values about backup.

**Comment by abarrak** [comment]
Hello @aoterolorenzo, 

Thanks. I've checked the goss failures and they were not related to this PR as they target pgpool and repmgr.
I may look into for anther PR to fix them.

**Comment by abarrak** [comment]
@aoterolorenzo I couldn't run the suite locally due cpu architecture for goss. 

```
$ k cp ./runtime-parameters.yaml postgres-postgresql-ha-pgpool-668f9cf54-krnnq:/tmp
$ k exec -it postgres-postgresql-ha-pgpool-668f9cf54-krnnq -- /tmp/goss-linux-amd64 --gossfile /tmp/goss.yaml --vars /tmp/runtime-parameters.yaml validate
OCI runtime exec failed: exec failed: unable to start container process: exec /tmp/goss-linux-amd64: exec format error: unknown
command terminated with exit code 126
$ k exec -it postgres-postgresql-ha-pgpool-668f9cf54-krnnq -- sh
$ uname -a
Linux postgres-postgresql-ha-pgpool-668f9cf54-krnnq 5.10.57 #1 SMP PREEMPT Fri Jul 14 22:49:12 UTC 2023 aarch64 GNU/Linux
$ dpkg --print-architecture
arm64
```

**However** can see from GH action all are passing and failure in `ginkgo` upload. 
Can you check and confirm if all is good?
https://github.com/bitnami/charts/actions/runs/6818469755/job/18544331131?pr=20563 
```
Pipeline result: failed
goss action: passed » Tests: 3 passed, 0 skipped, 0 failed
ginkgo action: failed » Tests: 0 passed, 0 skipped, 1 failed
goss action: passed » Tests: 8 passed, 0 skipped, 0 failed
Actions: 2 passed, 0 skipped, 1 failed, 3 total
```

**Comment by abarrak** [comment]
@aoterolorenzo Thanks, 

I can see from GH action all are passing and failure in `ginkgo` upload. 
https://github.com/bitnami/charts/actions/runs/6818469755/job/18544331131?pr=20563 

```
Pipeline result: failed
goss action: passed » Tests: 3 passed, 0 skipped, 0 failed
ginkgo action: failed » Tests: 0 passed, 0 skipped, 1 failed
goss action: passed » Tests: 8 passed, 0 skipped, 0 failed
Actions: 2 passed, 0 skipped, 1 failed, 3 total
```

With ginkgo, it was failing due the chart's password parameter is not passed correctly.

```bash
psql: error: connection to server at "postgres-postgresql-ha-pgpool" (10.105.64.76), port 5432 failed: ERROR:  authentication failed
DETAIL:  username "postgres" or password does not exist in backend
```

After setting the correct one:

```bash
bash-3.2$ ginkgo -- --kubeconfig=/Users/aalotai1/.kube/config --namespace=postgres --name=postgres-postgresql-ha --username=postgres --password=AgHtB1o6Zr
Ginkgo detected a version mismatch between the Ginkgo CLI and the version of Ginkgo imported by your packages:
  Ginkgo CLI Version:
    2.13.0
  Mismatched package versions found:
    2.11.0 used by ginkgo

  Ginkgo will continue to attempt to run but you may see errors (including flag
  parsing errors) and should either update your go.mod or your version of the
  Ginkgo CLI to match.

  To install the matching version of the CLI run
    go install github.com/onsi/ginkgo/v2/ginkgo
  from a path that contains a go.mod file.  Alternatively you can use
    go run github.com/onsi/ginkgo/v2/ginkgo
  from a path that contains a go.mod file to invoke the matching version of the
  Ginkgo CLI.

  If you are attempting to test multiple packages that each have a different
  version of the Ginkgo library with a single Ginkgo CLI that is currently
  unsupported.
  
Running Suite: PostgreSQL HA Persistence Test Suite - /Users/aalotai1/sre_home/code/bitnam_charts/.vib/postgresql-ha/ginkgo
===========================================================================================================================
Random Seed: 1699576247

Will run 1 of 1 specs

•

Ran 1 of 1 Specs in 166.094 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS
```

I seems this is [somehow because of the fixed password set here](https://github.com/bitnami/charts/pull/18752/files#diff-4abf04e9addbad30fbd2fc66ddeb167ff9318511384d9cad789e558663077ff0) by @javsalgar.

**Comment by aoterolorenzo** [comment]
> With ginkgo, it was failing due the chart's password parameter is not passed correctly.
> 
> ```shell
> psql: error: connection to server at "postgres-postgresql-ha-pgpool" (10.105.64.76), port 5432 failed: ERROR:  authentication failed

@abarrak is this error happening on the backup CronJob? Otherwise would be weird to see it

**Comment by abarrak** [comment]
@aoterolorenzo That's what I noticed and thought it was bash expansion issue and it's not.
As for the issue, it's not related to this PR, [failing here too.](https://github.com/bitnami/charts/pull/20936)

**Comment by aoterolorenzo** [comment]
Hi @abarrak, 

Shouldn't be happening again after merging with main. I will retry it anyway.

File: .vib/postgresql-ha/runtime-parameters.yaml [hunk 2]
```diff
@@ -12,11 +12,11 @@ pgpool:
     postgresql: 5432
 postgresql:
   repmgrUsername: repmgr
-  repmgrPassword: repmgrPassword123!4
+  repmgrPassword: repmgrPassword123@4
   repmgrDatabase: repmgr
   repmgrConnectTimeout: 7
   replicaCount: 4
-  password: psqlPassword123!4
+  password: psqlPassword123@4
```
Comment (Mauraza) [inline .vib/postgresql-ha/runtime-parameters.yaml:19]: Please, could you revert this change?

File: .vib/postgresql-ha/runtime-parameters.yaml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 pgpool:
   adminUsername: adminUser
-  adminPassword: adminPassword123!4
+  adminPassword: adminPassword123@4
```
Comment (Mauraza) [inline .vib/postgresql-ha/runtime-parameters.yaml:3]: Please, could you revert this change?

File: .vib/postgresql-ha/runtime-parameters.yaml [hunk 4]
```diff
@@ -12,11 +12,11 @@ pgpool:
     postgresql: 5432
 postgresql:
   repmgrUsername: repmgr
-  repmgrPassword: repmgrPassword123!4
+  repmgrPassword: repmgrPassword123@4
```
Comment (Mauraza) [inline .vib/postgresql-ha/runtime-parameters.yaml:15]: Could you revert this change?

**Review by Mauraza — CHANGES_REQUESTED** [review]
Hi @abarrak,

Thanks again for the contribution! Could you resolve the conflicts and check my comments?

File: .vib/postgresql-ha/runtime-parameters.yaml (hunk 2 above)
Comment (abarrak) [reply .vib/postgresql-ha/runtime-parameters.yaml]: reverted.

**Comment by abarrak** [comment]
@Mauraza integrations' pass is reverted.

I couldn't workaround the Markdown bot, it keeps splitting network args and patch last 3 to the new table.
https://github.com/bitnami/charts/pull/20563/commits/7f151c14932275937d54ac2d4e53dc09e2dba6db

**Comment by Mauraza** [comment]
Hi @abarrak,

You don't worry about the README.md because there is a GitHub action that runs the linter. Could you resolve the conflicts of  `bitnami/postgresql-ha/Chart.yaml`?

**Comment by abarrak** [comment]
@Mauraza conflict's resolved.

**Review by Mauraza — APPROVED** [review]
Hi @abarrak,

Thanks again for the contribution!! LGTM!!

**Review by aoterolorenzo — APPROVED** [review]

**Pull request merged**